C2DEVEL-14695: return backup service support

RudeZwloki7 opened the pull request:
New data sources:
- aws_backup_plan
- aws_backup_selection
- aws_backup_vault

New resources:
- aws_backup_plan
- aws_backup_selection
- aws_backup_vault_default

PixarV commented on internal/provider/provider.go:
```diff
@@ -16,9 +16,12 @@ import (
 	"github.com/hashicorp/terraform-provider-aws/internal/service/ec2"
 	"github.com/hashicorp/terraform-provider-aws/internal/service/eks"
 	"github.com/hashicorp/terraform-provider-aws/internal/service/elbv2"
+	// "github.com/hashicorp/terraform-provider-aws/internal/service/iam"
```
Лишние закомментированные импорты

RudeZwloki7 replied:
Убрал

PixarV commented on internal/provider/provider.go:
```diff
@@ -835,11 +838,11 @@ func Provider() *schema.Provider {
 			//"aws_backup_framework":                backup.ResourceFramework(),
 			//"aws_backup_global_settings":          backup.ResourceGlobalSettings(),
-			//"aws_backup_plan":                     backup.ResourcePlan(),
+			"aws_backup_plan":                     backup.ResourcePlan(),
```
Прогони `go fmt` на измененных файлах или `make fmt` для проекта, много где форматирование неправильное

RudeZwloki7 replied:
Прогнал

PixarV commented on internal/conns/config.go:
```diff
@@ -142,9 +142,20 @@ func (c *Config) Client(ctx context.Context) (interface{}, diag.Diagnostics) {
-	accountID, partition, err := awsbase.GetAwsAccountIDAndPartition(ctx, cfg, &awsbaseConfig)
-	if err != nil {
-		return nil, diag.Errorf("error retrieving account details: %s", err)
+	var accountID string
```
```suggestion
	var accountID, partition string
```

RudeZwloki7 replied:
Поправил

PixarV commented on internal/conns/config.go:
```diff
@@ -142,9 +142,20 @@ func (c *Config) Client(ctx context.Context) (interface{}, diag.Diagnostics) {
+
+	// FIXME: A temporary solution for building correct ARNs while yet resolving via IAM not implemented
+	if c.SkipCredsValidation && c.SkipRequestingAccountId {
+		var parsedAccountID = strings.Split(c.AccessKey, ":")
```
```suggestion
		parsedAccountID := strings.Split(c.AccessKey, ":")
```
Советует [styleguide](https://google.github.io/styleguide/go/best-practices.html#initialization)

И кажется, что `parsedAccountID` не совсем корректное название для этой переменной. Может `keyParts` или что-то такое

RudeZwloki7 replied:
Поправил, вариант `accessKeyParts` мне кажется норм

PixarV commented on internal/conns/config.go:
```diff
@@ -142,9 +142,20 @@ func (c *Config) Client(ctx context.Context) (interface{}, diag.Diagnostics) {
+		var parsedAccountID = strings.Split(c.AccessKey, ":")
+		accountID = parsedAccountID[len(parsedAccountID) - 1]
+		partition = "c2"
+		c.Region = ""
```
К сожалению, так все таки сделать нельзя. Некоторые датасорсы сохраняют значение из этого параметра в качестве id (например, `aws_availability_zones`), а пустое значение как id использовать нельзя. Так что от обнуления региона пока предлагаю отказаться. Обсудим

RudeZwloki7 replied:
Вместо обнуления региона в конфиге перенесли проставление пустого региона при построении ARN провайдером в бэкапируемых сущностях

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -41,48 +36,48 @@ The following arguments are supported:
 * `rule` - (Required) A rule object that specifies a scheduled task that is used to back up a selection of resources.
-* `advanced_backup_setting` - (Optional) An object that specifies backup options for each resource type.
-* `tags` - (Optional) Metadata that you can assign to help organize the plans you create. If configured with a provider [`default_tags` configuration block](/docs/providers/aws/index.html#default_tags-configuration-block) present, tags with matching keys will overwrite those defined at the provider-level.
+* `advanced_backup_setting` - (Optional) An object that specifies backup options for each resource type. _(Currently unsupported)_.
```
Неподдерживаемые параметры оформляются иначе. Как пример оформления доки можно использовать `ec2_transit_gateway.html.markdown`

Формулировки для описания параметров можно брать из нашей документации, обычно они более четко сфомулированы.

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -41,48 +36,48 @@ The following arguments are supported:
+* `backup_options` - (Required) Specifies the backup option for a selected resource. This option is only available for Windows VSS backup jobs. Set to `{ WindowsVSS = "enabled" }` to enable Windows VSS backup option and create a VSS Windows backup. _(Currently unsupported)_.
+* `resource_type` - (Required) The type of AWS resource to be backed up. For VSS Windows backups, the only supported resource type is Amazon EC2. Valid values: `EC2`. _(Currently unsupported)_.
 
 ## Attributes Reference
```
```suggestion
## Attribute Reference
```

Так для каждой странички

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+---
+subcategory: "Backup"
+layout: "aws"
+page_title: "AWS: aws_backup_selection"
```
```suggestion
page_title: "aws_backup_selection"
```

Так для всех страничек

RudeZwloki7 replied:
Убрал

PixarV commented on internal/service/backup/plan.go:
```diff
@@ -172,11 +172,18 @@ func resourcePlanCreate(d *schema.ResourceData, meta interface{}) error {
+	}
+
+	advancedBackupSettings := d.Get("advanced_backup_setting").(*schema.Set)
+	if len(advancedBackupSettings.List()) > 0 {
```
Есть метод `Len`, который тут можно использовать

У aws есть документация со стандартными вариантами маппинга из терраформ конфигурации в параметр апи и обратно. Их можно использовать как базу: https://github.com/C2Devel/terraform-provider-rockitcloud/blob/develop/docs/contributing/data-handling-and-conversion.md#root-typeset-of-resource-and-aws-list-of-structure

\+ похожие условия ниже

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/plan.go:
```diff
@@ -172,11 +172,18 @@ func resourcePlanCreate(d *schema.ResourceData, meta interface{}) error {
+		input.BackupPlan.AdvancedBackupSettings = expandBackupPlanAdvancedBackupSettings(advancedBackupSettings)
+	}
+
+	if len(tags) > 0 {
```
С тегами есть проблема: можно указать блок `default_tags` на уровне конфига провайдера, и тогда он будет пытаться добавить их для каждого ресурса, и это условие даст true. Предлагаю полностью выпилить работу с тегами из ресурса, включая элементы схемы

RudeZwloki7 replied:
Выпилил теги для бэкапа

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -20,18 +20,13 @@ resource "aws_backup_plan" "example" {
     rule_name         = "tf_example_backup_rule"
     target_vault_name = aws_backup_vault.test.name
     schedule          = "cron(0 12 * * ? *)"
```
Падает с ошибкой `InvalidParameterValue: Cron expression is not valid: There should be exactly 5 fields.`

RudeZwloki7 replied:
Поправил экзамплы, проверил локально

PixarV commented on website/docs/r/backup_vault.html.markdown:
```diff
@@ -10,12 +10,13 @@ description: |-
 
 Provides an AWS Backup vault resource.
 
+Deleting the vault is not supported in security purposes. In most cases, use the backup vault data source for existing default vault instead of the resource.
```
Это можно оформить в виде замечания:
```
~> <text>
```

Сейчас при дестрое конфигурации с волтом возникает ошибка. Если удаление не поддерживается, на дестрое как вариант можно печатать варнинг и удалять ресурс из состояния.

RudeZwloki7 replied:
Оформил

PixarV commented on internal/service/backup/plan.go:
```diff
@@ -246,12 +255,16 @@ func resourcePlanUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 
+		advancedBackupSettings := d.Get("advanced_backup_setting").(*schema.Set)
+		if len(advancedBackupSettings.List()) > 0 {
```
Len + https://github.com/C2Devel/terraform-provider-rockitcloud/blob/develop/docs/contributing/data-handling-and-conversion.md#root-typeset-of-resource-and-aws-list-of-structure

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -184,14 +185,26 @@ func resourceSelectionCreate(d *schema.ResourceData, meta interface{}) error {
+	}
+
+	notResources := d.Get("not_resources").(*schema.Set)
+	if len(notResources.List()) > 0{
```
Len + https://github.com/C2Devel/terraform-provider-rockitcloud/blob/develop/docs/contributing/data-handling-and-conversion.md#root-typeset-of-resource-and-aws-list-of-structure

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -184,14 +185,26 @@ func resourceSelectionCreate(d *schema.ResourceData, meta interface{}) error {
 	}
 
+	conditions := d.Get("condition").(*schema.Set).List()
+	if len(conditions)>0{
```
см https://github.com/C2Devel/terraform-provider-rockitcloud/blob/develop/docs/contributing/data-handling-and-conversion.md#root-typeset-of-resource-and-aws-list-of-structure

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -334,6 +347,47 @@ func resourceSelectionRead(d *schema.ResourceData, meta interface{}) error {
+	if d.HasChanges("name", "resources") {
+		selection := &backup.Selection{
+			IamRoleArn:    aws.String(d.Get("iam_role_arn").(string)),
+			Resources:     flex.ExpandStringSet(d.Get("resources").(*schema.Set)),
```
Если `resources` - это обновляемый атрибут, нужно убрать из схемы `ForceNew: true` + сейчас он там как опциональный отмечен, а в доке - обязательный

RudeZwloki7 replied:
Поправил, сделал `resources` обязательным полем

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -334,6 +347,47 @@ func resourceSelectionRead(d *schema.ResourceData, meta interface{}) error {
+		if err != nil {
+			return fmt.Errorf("error updating Backup Selection: %s", err)
+		}
+		d.SetId(aws.StringValue(output.SelectionId))
```
После апдейта по идее не должно требоваться сетить id. `SelectionId` поменяется?

RudeZwloki7 replied:
Это лишнее, убрал

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -334,6 +347,47 @@ func resourceSelectionRead(d *schema.ResourceData, meta interface{}) error {
+func resourceSelectionUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*conns.AWSClient).BackupConn
+
+	if d.HasChanges("name", "resources") {
```
`"iam_role_arn"` тоже можно сюда добавить, раз уж его включаем в запрос

RudeZwloki7 replied:
Решили, что `iam_role_arn` будет computed полем

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -41,48 +36,48 @@ The following arguments are supported:
 ### Rule Arguments
 For **rule** the following attributes are supported:
 
 * `rule_name` - (Required) An display name for a backup rule.
```
У поля есть валидации в схеме на длину и регехп. Они актуальны для бэкап апи (в доке инфы нет)?
Если актуальны, надо добавить инфу о них в документацию.

Аналогично для  `target_vault_name`. Там они вообще кажется не нужны, тк это имя существующего волта.

RudeZwloki7 replied:
Добавил инфу по валидациям

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -334,6 +347,47 @@ func resourceSelectionRead(d *schema.ResourceData, meta interface{}) error {
+			output, err = conn.UpdateBackupSelection(input)
+
+			if err != nil {
+				return resource.NonRetryableError(err)
```
Если любая ошибка - NonRetryable, то сама конструкция retry не нужна. Можно заменить весь блок на что-то типа:
```
output, err := conn.UpdateBackupSelection(input)
if err != nil {
    return fmt.Errorf("error updating Backup Selection: %s", err)
}
```

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -334,6 +347,47 @@ func resourceSelectionRead(d *schema.ResourceData, meta interface{}) error {
+			SelectionId:     aws.String(d.Id()),
+			BackupSelection: selection,
+		}
+
```
Добавь пожалуйста логирование получившегося инпута

RudeZwloki7 replied:
Добавил

PixarV commented on website/docs/d/backup_selection.html.markdown:
```diff
@@ -31,5 +31,5 @@ The following arguments are supported:
 In addition to all arguments above, the following attributes are available:
 
 * `name` - The display name of a resource selection document.
```
Проверить валидации

RudeZwloki7 replied:
Добавил инфу по валидациям

PixarV commented on internal/service/backup/vault.go:
```diff
@@ -64,9 +64,11 @@ func resourceVaultCreate(d *schema.ResourceData, meta interface{}) error {
-		BackupVaultTags: Tags(tags.IgnoreAWS()),
 	}
 
+	if len(tags) > 0 {
```
Теги лучше совсем выпилить

RudeZwloki7 replied:
Убрал в `vault_default`

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -41,48 +36,48 @@ The following arguments are supported:
 * `schedule` - (Optional) A CRON expression specifying when AWS Backup initiates a backup job.
-* `enable_continuous_backup` - (Optional) Enable continuous backups for supported resources.
+* `enable_continuous_backup` - (Optional) Enable continuous backups for supported resources. _(Currently unsupported)_.
 * `start_window` - (Optional) The amount of time in minutes before beginning a backup.
```
Для поддерживаемых **аргументов**, которые имеют дефолтные значения в схеме, в конце описания добаляется `Defaults to <value>.`

RudeZwloki7 replied:
Добавил

PixarV commented on website/docs/r/backup_vault.html.markdown:
```diff
@@ -24,8 +25,8 @@ resource "aws_backup_vault" "example" {
 The following arguments are supported:
 
 * `name` - (Required) Name of the backup vault to create.
```
Проверить валидации

RudeZwloki7 replied:
Убрали `name` из аргументов в соответствии с текущим поведением

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+# Resource: aws_backup_selection
+
+Manages selection conditions for AWS Backup plan resources.
```
```suggestion
Manages selection conditions for backup plan resources.
```
\+ продублировать эту строку в тег `description` выше

Если имеется ввиду сервис, то с большой буквы. Если имеется ввиду ресурс (тут backup plan), то с маленькой.

PixarV replied:
Стараюсь использовать одинаковые формулировки в описании для ресурсов и датасорсов:
- для ресурса: `Manages a/an <resource>. For details about <service>, see the [user documentation](ссылка на облачную доку, если есть страница для сервиса).`
- для датасорса: `Provides information about a/an <resource>.`

В качестве примера смотри страницы для paas_\*, ec2_transit_gateway_\*

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_selection.html.markdown:
````diff
@@ -0,0 +1,61 @@
+
+```terraform
+resource "aws_backup_selection" "example" {
+  iam_role_arn = ""
````
Это неподдерживаемый параметр (по доке), можно убрать совсем

RudeZwloki7 replied:
Убрал, сделали computed полем

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+resource "aws_backup_selection" "example" {
+  iam_role_arn = ""
+  name         = "tf_example_backup_selection"
+  plan_id      = aws_backup_plan.example.id
```
Если твой ресурс ссылается на другой ресурс, желательно приложить конфигурацию и для него тоже в пример, чтобы его можно было скопировать и запустить целиком.

RudeZwloki7 replied:
Добавил

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+  name         = "tf_example_backup_selection"
+  plan_id      = aws_backup_plan.example.id
+
+  resources = [
```
Кажется из этого списка мы поддерживаем только `aws_ebs_volume.example.arn`. Пример в доке должен быть рабочий + нужна конфигурация для указанных ресурсов

RudeZwloki7 replied:
Протестил еще с инстансом, но решил не пихать, чтобы не увеличивать и так объемный пример. Убрал все кроме ebs volume

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+The following arguments are supported:
+
+* `name` - (Required) The display name of a resource selection document.
```
Для описания полей лучше брать версию из нашей документации, которая обычно более лаконична и однозначна. Также сюда добавляется информация о валидациях и дефолтных значениях для полей, возможных значениях если они есть. В качестве примера формулировок смотри страницы для paas_\*, ec2_transit_gateway_\*

Для редактируемых полей можно добавить `Editable` в начало. Например, `(Required, Editable)`

RudeZwloki7 replied:
Добавил

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+* `name` - (Required) The display name of a resource selection document.
+* `plan_id` - (Required) The backup plan ID to be associated with the selection of resources.
+* `iam_role_arn` - (Required) The ARN of the IAM role that AWS Backup uses to authenticate when restoring and backing up the target resource. See the [AWS Backup Developer Guide](https://docs.aws.amazon.com/aws-backup/latest/devguide/access-control.html#managed-policies) for additional information about using AWS managed policies or creating custom policies attached to the IAM role. _(Currently u ...
```
Изменить оформление неподдерживаемых полей

При переносе полей в неподдерживаемые атрибуты можно срезать часть описания до первого предложения + ссылки могут быть для нас не актуальны

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+* `key` - (Required) The key in a key-value pair. _(Currently unsupported)_.
+* `value` - (Required) The value in a key-value pair. _(Currently unsupported)_.
+
+## Attributes Reference
```
```suggestion
## Attribute Reference
```

Так для всех

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+## Import
+
+Backup selection can be imported using the role plan_id and id separated by `|`.
```
```suggestion
Backup selection can be imported using the `plan_id` and `id` separated by `|`.
```

RudeZwloki7 replied:
Убрал блок импорта в этом месте, не сработал

RudeZwloki7 replied:
Вернул

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - Backup Selection identifier
```
```suggestion
* `id` - The ID of the backup selection.
```

RudeZwloki7 replied:
Поправил, взял строку из доки

PixarV commented on website/docs/r/backup_vault.html.markdown:
```diff
@@ -10,12 +10,13 @@ description: |-
 
 Provides an AWS Backup vault resource.
```
```suggestion
Manages a backup vault.
```

RudeZwloki7 replied:
поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -20,18 +20,13 @@ resource "aws_backup_plan" "example" {
     rule_name         = "tf_example_backup_rule"
     target_vault_name = aws_backup_vault.test.name
     schedule          = "cron(0 12 * * ? *)"
+    start_window             = 60
```
Прогони таргет ` make website-lint-fix`, он поправит ошибки линтинга (тут выровняет конфигурацию)

RudeZwloki7 replied:
поправил

PixarV commented on website/docs/r/backup_vault.html.markdown:
````diff
@@ -10,12 +10,13 @@ description: |-
 
 ```terraform
 resource "aws_backup_vault" "example" {
   name        = "example_backup_vault"
````
В примерах в имя лучше добавлять инфу про терраформ, чтобы было понятно откуда ресурс.

Я обычно использую что-то типа `tf-backup-vault`

RudeZwloki7 replied:
поправил

PixarV reviewed (changes requested):
Нужно перепроверить коды ошибок, которые проверяются в некоторых ифах: `backup.ErrCodeResourceNotFoundException`, `backup.ErrCodeInvalidRequestException` и пр

Гитхаб не позволяет удобно писать комменты к необновленному коду, так что комментарии к странице `backup_selection.html.markdown` относятся на самом деле к каждой страничке)

RudeZwloki7 committed:
ec2: remove region from ARN for ebs volume and instance

RudeZwloki7 committed:
backup: add default vault resource, update vault data source

RudeZwloki7 committed:
backup: update backup plan resource and data source

RudeZwloki7 committed:
backup: update backup selection resource

RudeZwloki7 committed:
provider: add backup data sources and resources to provider schema

Add backup plan, backup selection, backup vault data sources and
resources to provider schema.

RudeZwloki7 committed:
config: set correct account id and partition for generated ARNs

RudeZwloki7 committed:
website: add backup vault data source and resource docs

RudeZwloki7 committed:
website: add backup plan data source and resource docs

RudeZwloki7 committed:
website: add backup selection data source and resource docs

RudeZwloki7 commented:
Обновил коды ошибок

PixarV commented on internal/provider/provider.go:
```diff
@@ -835,11 +836,11 @@ func Provider() *schema.Provider {
-			//"aws_backup_selection":                backup.ResourceSelection(),
-			//"aws_backup_vault":                    backup.ResourceVault(),
+			"aws_backup_selection": backup.ResourceSelection(),
+			"aws_backup_vault":     backup.ResourceDefaultVault(),
```
Я бы добавила совсем новый ресурс `aws_backup_default_vault`, а `aws_backup_vault` оставила закомментированным. Если волты начнут поддерживать CRUD полностью, можно будет открыть доступ к `aws_backup_vault` и точно ничего не поломать

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/backup"
+
```
Прогони `make importlint`, по-моему эта пустая строка лишняя

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/plan.go:
```diff
@@ -246,26 +225,22 @@ func resourcePlanUpdate(d *schema.ResourceData, meta interface{}) error {
-				BackupPlanName:         aws.String(d.Get("name").(string)),
-				Rules:                  expandBackupPlanRules(d.Get("rule").(*schema.Set)),
-				AdvancedBackupSettings: expandBackupPlanAdvancedBackupSettings(d.Get("advanced_backup_setting").(*schema.Set)),
+				BackupPlanName: aws.String(d.Get("name").(string)),
```
Тк у нас можно обновлять имя плана, можно удалить опцию `ForceNew: true` для параметра `name` в схеме + добавить это поле выше в `if d.HasChanges(..)`

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -183,15 +178,26 @@ func ResourceSelection() *schema.Resource {
 func resourceSelectionCreate(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*conns.AWSClient).BackupConn
 
+	d.Set("iam_role_arn", "")
```
Немного странно встретить тут `d.Set(..)`

Я бы может сразу указала бы "" в реквесте с поясняющим комментом:
```
	selection := &backup.Selection{
		IamRoleArn: aws.String(""), // FIXME: add explanation
        }
```
А в `resourceSelectionRead` уже засетится значение

RudeZwloki7 replied:
Поправил проставление значения в запросе + убрал проставление значения в `Read`

PixarV commented on internal/service/backup/selection.go:
```diff
@@ -334,6 +340,38 @@ func resourceSelectionRead(d *schema.ResourceData, meta interface{}) error {
+
+		_, err := conn.UpdateBackupSelection(input)
+
+		if tfresource.TimedOut(err) {
```
Вот этот блок можно убрать, он работает с ошибкой, которую возвращает `resource.Retry`  и схлопнуть вызов Update и проверку ошибки до:
```
if _, err := conn.UpdateBackupSelection(input); err != nil {
	return fmt.Errorf("error updating Backup Selection: %s", err)
}
```

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/d/backup_plan.html.markdown:
```diff
@@ -1,9 +1,9 @@
+page_title: "aws_backup_plan"
 description: |-
-  Provides details about an AWS Backup plan.
+  Provides details about an backup plan.
```
```suggestion
  Provides information about a backup plan.
```

Эта строка дублируется в описание под именем ресурса 

```
# Data Source: aws_backup_plan

Provides information about a backup plan.
```

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"kms_key_arn": {
```
В ресурсы, которые мы сами добавляем, не нужно включать неподдерживаемые поля.

RudeZwloki7 replied:
Убрал `kms_key_arn` из ресурса

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"recovery_points": {
```
Тоже не поддерживается

RudeZwloki7 replied:
recovery points поддерживаются, апиха отдает данные

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+		}
+
+		d.SetId(name)
+	}
```
Не обрабатывается условие, когда `ListBackupVaults` вернул ошибку, которая != nil и ее код не из `errCodeVaultNotFound`

RudeZwloki7 replied:
Добавил обработку ошибки

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+		log.Printf("[INFO] Found existing Backup Vault (%s)", *vault.BackupVaultName)
+
+		d.SetId(*vault.BackupVaultName)
+	} else if tfawserr.ErrCodeEquals(err, errCodeVaultNotFound) || len(output.BackupVaultList) == 0 {
```
`ListBackupVaults` без параметров может вернуть код из `errCodeVaultNotFound`?

RudeZwloki7 replied:
Нет, вернет пустой список. Убрал `errCodeVaultNotFound`

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+			input.EncryptionKeyArn = aws.String(v.(string))
+		}
+
+		_, err = conn.CreateBackupVault(input)
```
Можно схлопнуть с проверкой ошибки в один if

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+
+	if err == nil && len(output.BackupVaultList) > 0 {
+		vault := output.BackupVaultList[0]
+		log.Printf("[INFO] Found existing Backup Vault (%s)", *vault.BackupVaultName)
```
```suggestion
		log.Printf("[INFO] Found existing Backup Vault (%s)", aws.StringValue(vault.BackupVaultName))
```

Для разыменования обычно используются утильные функции

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+		vault := output.BackupVaultList[0]
+		log.Printf("[INFO] Found existing Backup Vault (%s)", *vault.BackupVaultName)
+
+		d.SetId(*vault.BackupVaultName)
```
`aws.StringValue(..)`

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/backup/vault_default.go:
```diff
@@ -0,0 +1,106 @@
+	input := &backup.ListBackupVaultsInput{}
+	output, err := conn.ListBackupVaults(input)
+
+	if err == nil && len(output.BackupVaultList) > 0 {
```
Предлагаю немного реорганизовать условия, чтобы уменьшить if блоки:

```
if err != nil {
    if tfawserr.ErrCodeEquals(err, errCodeVaultNotFound) || len(output.BackupVaultList) == 0 {
        vault := output.BackupVaultList[0]
        ...
        return resourceDefaultVaultRead(d, meta)
    } else {
        return fmt.Errorf("err explanation")
    }
}

// create logic
...
return resourceDefaultVaultRead(d, meta)
```

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/d/backup_plan.html.markdown:
```diff
@@ -24,11 +24,15 @@ The following arguments are supported:
+->  **Unsupported attributes**
+These attributes are currently unsupported:
+
+* `tags` - Metadata that you can assign to help organize the plans you create. Excluded from schema.
```
Можно убрать, тк тэгов больше нет

RudeZwloki7 replied:
Убрал выпиленные теги из доки по обновленным ресурсам

PixarV commented on website/docs/d/backup_plan.html.markdown:
````diff
@@ -14,7 +14,7 @@ Use this data source to get information on an existing backup plan.
 ```terraform
 data "aws_backup_plan" "example" {
-  plan_id = "tf_example_backup_plan_id"
+  plan_id = "tf-backup-plan-id"
````
А в каком формате обычно plan_id? если это uuid, то лучше использовать юидообразный шаблон

RudeZwloki7 replied:
Поправил в примерах id на шаблон `01234567-0123-0123-0123-0123456789ab`

PixarV commented on website/docs/d/backup_selection.html.markdown:
```diff
@@ -0,0 +1,43 @@
+layout: "aws"
+page_title: "aws_backup_selection"
+description: |-
+  Provides details about an backup selection.
```
```suggestion
  Provides details about a backup selection.
```

RudeZwloki7 replied:
Поправил

PixarV commented on internal/service/ec2/ebs_volume.go:
```diff
@@ -192,7 +192,7 @@ func resourceEBSVolumeRead(d *schema.ResourceData, meta interface{}) error {
 		Partition: meta.(*conns.AWSClient).Partition,
 		Service:   ec2.ServiceName,
-		Region:    meta.(*conns.AWSClient).Region,
+		Region:    "",
```
Добавь пожалуйста коммент о том, что регион в арнах не поддерживается

RudeZwloki7 replied:
Поправил в volume и instance ресурсах и датасорсах

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,97 @@
+* `name` - (Required) The name of the backup plan.
+* `rule` - (Required, Editable) A rule object that specifies a separate periodic task.
+
+->  **Unsupported arguments**
```
Я их обычно в раздел `Unsupported attributes` перемещала, тк формально они тут описываются только потому что они есть в итоговом состоянии. И получался один такой раздел на одну страничку в документации

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_vault.html.markdown:
```diff
@@ -0,0 +1,54 @@
+layout: "aws"
+page_title: "aws_backup_vault"
+description: |-
+  Manages an backup vault resource.
```
```suggestion
  Manages a backup vault resource.
```

RudeZwloki7 replied:
Поправил

PixarV reviewed (changes requested):
Документацию еще не досмотрела

Переоформи еще пожалуйста немного описание ПРа. Пример можешь посмотреть тут: #32

PixarV commented on website/docs/d/backup_plan.html.markdown:
```diff
@@ -0,0 +1,37 @@
+
+# Data Source: aws_backup_plan
+
+  Provides information about a backup plan.
```
```suggestion
Provides information about a backup plan.
```

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/d/backup_selection.html.markdown:
```diff
@@ -0,0 +1,43 @@
+
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are available:
```
```suggestion
In addition to all arguments above, the following attributes are exported:
```

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+  Manages a backup plan resource.
+---
+
+[Backup]: https://docs.cloud.croc.ru/en/services/backup/operations.html#backupplan
```
```suggestion
[backup-plan]: https://docs.cloud.croc.ru/en/services/backup/operations.html#backupplan
```

На случай, если еще ссылки из этого раздела появятся ) + обновить в месте использования

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+The following arguments are supported:
+
+* `name` - (Required) The name of the backup plan.
+* `rule` - (Required, Editable) A rule object that specifies a separate periodic task.
```
Тут на самом деле список объектов

В конфигурации будет указываться так:
```
resource "aws_backup_plan" "example" {
  name = "tf-backup-plan"
  
  rule {...}
  
  rule {...}
}
```

PixarV replied:
Для перехода к описанию внутренней структуры блока добавь в конец предложение: `The structure of the block is [described below](#rule).`. 

`#rule` - это ссылка на соответствующий заголовок внизу, там будет `### Rule` вместо `### Rule Arguments`

RudeZwloki7 replied:
Добавил еще один блок `rule` в примерах конфигураций + в описании поля `rule`

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+* `rule` - (Required, Editable) A rule object that specifies a separate periodic task.
+
+### Rule Arguments
+For **rule** the following arguments are supported:
```
Используй тут такую формулировку: ``The `rule` block has the following structure:``

Еще до этого предложения можно добавить такой блок, если все параметры редактируемые:
``~> All the parameters in the `rule` block are editable.``

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+For **rule** the following arguments are supported:
+
+* `completion_window` - (Optional) Time in minutes after the start of the backup job, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.
+* `lifecycle` - (Optional) The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. Fields documented below.
```
```suggestion
* `lifecycle` - (Optional) The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. The structure of the block is [described below](#lifecycle).
```

Правки как для `rule`

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+
+* `completion_window` - (Optional) Time in minutes after the start of the backup job, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.
+* `lifecycle` - (Optional) The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. Fields documented below.
+* `rule_name` - (Required) The name of the backup rule. Must be at least `1` character and at most `50` characters in length. Must contain only alphanumeric characters, hyphens, underscores, and periods.
```
```suggestion
* `rule_name` - (Required) The name of the backup rule. The value must be 1 to 50 characters long and must contain only alphanumeric characters, hyphens, underscores, and periods.
```

Так чуть короче

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the backup plan.
```
Тожe Editable

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+* `rule_name` - (Required) The name of the backup rule. Must be at least `1` character and at most `50` characters in length. Must contain only alphanumeric characters, hyphens, underscores, and periods.
+* `schedule` - (Optional) CRON expression in UTC for the backup job scheduling.
+* `start_window` - (Optional) Time in minutes during which a backup job should start running. Otherwise the job will be cancelled. Defaults to `60`.
+* `target_vault_name` - (Required) The name of the backup vault. Must be at least `2` characters and at most `50` characters in length. Must contain only alphanumeric characters, hyphens, and underscores.
```
```suggestion
* `target_vault_name` - (Required) The name of the backup vault. The value must be 2 to 50 characters long and must contain only alphanumeric characters, hyphens, and underscores.
```

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+* `start_window` - (Optional) Time in minutes during which a backup job should start running. Otherwise the job will be cancelled. Defaults to `60`.
+* `target_vault_name` - (Required) The name of the backup vault. Must be at least `2` characters and at most `50` characters in length. Must contain only alphanumeric characters, hyphens, and underscores.
+
+### Lifecycle Arguments
```
см. комментарии для rule

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_vault_default.html.markdown:
```diff
@@ -0,0 +1,47 @@
+* `name` - The name of the backup vault.
+* `recovery_points` - The number of recovery points in the vault.
+
+->  **Unsupported attributes**
```
Этот блок тут не нужен

RudeZwloki7 replied:
Убрал

PixarV commented on website/docs/r/backup_vault_default.html.markdown:
````diff
@@ -0,0 +1,47 @@
+## Example Usage
+
+```terraform
+resource "aws_backup_vault_default" "example" {
````
```suggestion
resource "aws_backup_vault_default" "example" {}
```

Указать в одну строку немного компактнее. \+ поправить в остальных примерах.

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_vault_default.html.markdown:
```diff
@@ -0,0 +1,47 @@
+* `arn` - The Amazon Resource Name (ARN) of the backup vault.
+* `id` - The name of the backup vault.
+* `name` - The name of the backup vault.
+* `recovery_points` - The number of recovery points in the vault.
```
```suggestion
* `recovery_points` - The number of recovery points in the backup vault.
```

Раз уж везде выше `the backup vault`

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+
+resource "aws_ebs_volume" "example" {
+  availability_zone = data.aws_availability_zones.this.names[0]
+  type              = "standard"
```
возьми тут `gp2` лучше

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_vault_default.html.markdown:
```diff
@@ -0,0 +1,47 @@
+
+## Import
+
+Backup vault can be imported using the `name`, e.g.,
```
```suggestion
Backup vault can be imported using `name`, e.g.,
```

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+
+## Import
+
+Backup Plan can be imported using the `id`, e.g.,
```
```suggestion
Backup Plan can be imported using `id`, e.g.,
```

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+
+The following arguments are supported:
+
+* `name` - (Required, Editable) The selection name. Must be at least `1` character and at most `50` characters in length. Must contain contain only alphanumeric, hyphen, underscore, and period characters.
```
```suggestion
* `name` - (Required, Editable) The selection name. The value must be 1 to 50 characters long and must contain contain only alphanumeric, hyphen, underscore, and period characters.
```

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+->  **Unsupported attributes**
+These attributes are currently unsupported:
+
+* `condition` - A list of conditions that you define to assign resources to your backup plans using tags.
```
`Always empty` вероятно потерял + ниже

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_vault_default.html.markdown:
```diff
@@ -0,0 +1,47 @@
+Manages a backup vault resource.
+Creates new default backup vault or uses existing backup vault as resource.
+
+~> Deleting the vault is not supported in security purposes.
```
Есть такая формулировка: ``~> Terraform does not delete the default backup vault when running `terraform destroy`, it will only be removed from the Terraform state.``

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_vault_default.html.markdown:
```diff
@@ -0,0 +1,47 @@
+
+# Resource: aws_backup_vault_default
+
+Manages a backup vault resource.
```
Я бы переписала эту пару строк как-нибудь так:
```
Manages the default backup vault. If it doesn't exist, Terraform will create one.
```

\+ поправить в description

RudeZwloki7 replied:
Поправил

PixarV commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,85 @@
+
+# Resource: aws_backup_plan
+
+Manages a backup plan resource. For details about backup plans, see the [user documentation][Backup].
```
```suggestion
Manages a backup plan. For details about backup plans, see the [user documentation][Backup].
```

RudeZwloki7 replied:
Поправил

DiGa66 commented on website/docs/d/backup_plan.html.markdown:
```diff
@@ -0,0 +1,37 @@
+->  **Unsupported attributes**
+These attributes are currently unsupported:
+
+* `version` - Unique, randomly generated, Unicode, UTF-8 encoded string that serves as the version ID of the backup plan. Always `""`.
```
that serves as -> used as

DiGa66 commented on website/docs/d/backup_selection.html.markdown:
```diff
@@ -0,0 +1,43 @@
+->  **Unsupported attributes**
+These attributes are currently unsupported:
+
+* `iam_role_arn` - The Amazon Resource Name (ARN) of the role used for authentication when the resource is backed up (not implemented, added to ensure compatibility with AWS). Always `""`.
```
надо ли писать про то, что не реализовано, если раздел называется "Неподдерживаемые атрибуты)
я бы убрал весь текст в скобках

RudeZwloki7 replied:
Конкретно по этому полю не реализована бизнес логика, но оно все еще обязательное в схеме и бэк проверяет его наличие в запросе. Сейчас подставляем дефолтное значение на уровне провайдера, чтобы облегчить жизнь пользователю. Думаю, совсем убирать не стоит, т.к. поле используется. Может просто описание поправить на более понятное?

DiGa66 commented on website/docs/d/backup_selection.html.markdown:
```diff
@@ -0,0 +1,43 @@
+In addition to all arguments above, the following attributes are exported:
+
+* `name` - The selection name.
+* `resources` - The list of Amazon Resource Names (ARNs) resources included into the backup plan.
```
of the resources

DiGa66 commented on website/docs/d/backup_selection.html.markdown:
```diff
@@ -0,0 +1,43 @@
+
+The following arguments are supported:
+
+* `plan_id` - (Required) The backup plan ID.
```
для единообразия тогда уж
The ID of the backup plan

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+The following arguments are supported:
+
+* `name` - (Required, Editable) The name of the backup plan.
+* `rule` - (Required, Editable) List of rule objects for separate periodic tasks. The structure of this block is [described below](#rule)
```
может просто rules ? в следующем предложении пишем, что это блок

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+
+The `rule` block has the following structure:
+
+* `completion_window` - (Optional) Time in minutes after the start of the backup job, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.
```
по красоте after the backup job is started, но необ.

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+The `rule` block has the following structure:
+
+* `completion_window` - (Optional) Time in minutes after the start of the backup job, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.
+* `lifecycle` - (Optional) The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. The structure of the block is [described below](#lifecycle).
```
transitioned скорее трансформация
tranferred

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+
+* `completion_window` - (Optional) Time in minutes after the start of the backup job, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.
+* `lifecycle` - (Optional) The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. The structure of the block is [described below](#lifecycle).
+* `rule_name` - (Required) The name of the backup rule. The value must be `1` to `50` characters long and must contain only alphanumeric characters, hyphens, underscores, and periods.
```
and -> or

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+
+The `lifecycle` block has the following structure:
+
+* `delete_after` - (Optional, Editable) Number of days from the recovery point creation date, after which it will be deleted.
```
the date of creation of the restore point

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+* `rule` - A rule object that specifies a separate periodic task.
+    * `copy_action` - Configuration block(s) with copy operation settings.
+    * `enable_continuous_backup` - Enable continuous backups for supported resources. Defaults to `false`.
+    * `lifecycle` - The lifecycle defines when a protected resource is transitioned to cold storage and when it expires.
```
Defines ...
transferred

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+    * `copy_action` - Configuration block(s) with copy operation settings.
+    * `enable_continuous_backup` - Enable continuous backups for supported resources. Defaults to `false`.
+    * `lifecycle` - The lifecycle defines when a protected resource is transitioned to cold storage and when it expires.
+        * `cold_storage_after` - Specifies the number of days after creation that a recovery point is moved to cold storage.
```
after creation of the recovery point after which it

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+    * `lifecycle` - The lifecycle defines when a protected resource is transitioned to cold storage and when it expires.
+        * `cold_storage_after` - Specifies the number of days after creation that a recovery point is moved to cold storage.
+    * `recovery_point_tags` - Metadata that you can assign to help organize the resources that you create.
+* `version` - Unique, randomly generated, Unicode, UTF-8 encoded string that serves as the version ID of the backup plan. Always `""`.
```
that servers -> used

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+    * `enable_continuous_backup` - Enable continuous backups for supported resources. Defaults to `false`.
+    * `lifecycle` - The lifecycle defines when a protected resource is transitioned to cold storage and when it expires.
+        * `cold_storage_after` - Specifies the number of days after creation that a recovery point is moved to cold storage.
+    * `recovery_point_tags` - Metadata that you can assign to help organize the resources that you create.
```
не оч. понятно, что значит organize в этом контексте

RudeZwloki7 replied:
У aws по тегам много где такая формулировка фигурирует. Как я понял, они подразумевают под organize управление в плане поиска, фильтрации, группировки ресурсов по их тегам

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+In addition to all arguments above, the following attributes are exported:
+
+* `arn` - The Amazon Resource Name (ARN) of the backup plan.
+* `id` - The backup plan ID.
```
см. выше

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+* `advanced_backup_setting` - An object that specifies backup options for each resource type. Always empty.
+* `rule` - A rule object that specifies a separate periodic task.
+    * `copy_action` - Configuration block(s) with copy operation settings.
+    * `enable_continuous_backup` - Enable continuous backups for supported resources. Defaults to `false`.
```
Enables

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+->  **Unsupported attributes**
+These attributes are currently unsupported:
+
+* `advanced_backup_setting` - An object that specifies backup options for each resource type. Always empty.
```
Specifies ... (начало убрать)

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+These attributes are currently unsupported:
+
+* `advanced_backup_setting` - An object that specifies backup options for each resource type. Always empty.
+* `rule` - A rule object that specifies a separate periodic task.
```
Specifies ...

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+
+## Import
+
+Backup Plan can be imported using `id`, e.g.,
```
plan

DiGa66 commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,96 @@
+
+The following arguments are supported:
+
+* `name` - (Required, Editable) The selection name. The value must be `1` to `50` characters long and must contain only alphanumeric, hyphen, underscore, and period characters.
```
лучше: 
alphanumeric characters, hyphens, underscores, and periods.
и в других местах поменять

DiGa66 commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,96 @@
+The following arguments are supported:
+
+* `name` - (Required, Editable) The selection name. The value must be `1` to `50` characters long and must contain only alphanumeric, hyphen, underscore, and period characters.
+* `plan_id` - (Required) The backup plan ID.
```
ID of ...

DiGa66 commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,96 @@
+
+* `name` - (Required, Editable) The selection name. The value must be `1` to `50` characters long and must contain only alphanumeric, hyphen, underscore, and period characters.
+* `plan_id` - (Required) The backup plan ID.
+* `resources` - (Required, Editable) The list of Amazon Resource Names (ARNs) resources included into the backup plan.
```
of the resources

DiGa66 commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,96 @@
+These attributes are currently unsupported:
+
+* `condition` - A list of conditions that you define to assign resources to your backup plans using tags. Always empty.
+* `iam_role_arn` -  The Amazon Resource Name (ARN) of the role used for authentication when the resource is backed up (not implemented, added to ensure compatibility with AWS). Always `""`.
```
нужен текст в скобках?

DiGa66 commented on website/docs/r/backup_selection.html.markdown:
```diff
@@ -0,0 +1,96 @@
+
+* `condition` - A list of conditions that you define to assign resources to your backup plans using tags. Always empty.
+* `iam_role_arn` -  The Amazon Resource Name (ARN) of the role used for authentication when the resource is backed up (not implemented, added to ensure compatibility with AWS). Always `""`.
+* `not_resources` - An array of strings that either contain Amazon Resource Names (ARNs) or match patterns of resources to exclude from a backup plan. Always empty.
```
массив же содержит? тогда
contains
matches

that contains either Amazon Resource Names (ARNs) or match patterns of resources to be excluded from a backup plan. Always empty.

DiGa66 commented on website/docs/r/backup_vault_default.html.markdown:
```diff
@@ -0,0 +1,41 @@
+In addition to all arguments above, the following attributes are exported:
+
+* `arn` - The Amazon Resource Name (ARN) of the backup vault.
+* `id` - The name of the backup vault.
```
The ID

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+
+The `rule` block has the following structure:
+
+* `completion_window` - (Optional) Time in minutes after the backup job is started, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.
```
cancelled

RudeZwloki7 replied:
Поправил

DiGa66 commented on website/docs/r/backup_plan.html.markdown:
```diff
@@ -0,0 +1,100 @@
+* `rule` - Specifies a separate periodic task.
+    * `copy_action` - Configuration block(s) with copy operation settings.
+    * `enable_continuous_backup` - Enables continuous backups for supported resources. Defaults to `false`.
+    * `lifecycle` - Defines when a protected resource is transferred to cold storage and when it expires.
```
а тут все правильно? Как сам экземпляр (защищаемый ресурс) может быть перемещен в cold storage?

RudeZwloki7 replied:
Т.к. этот атрибут неподдерживаемый, за основу взята формулировка от aws. По идее такие формулировки не особо критичны, пока у нас нет нашей реализации, под которую можно четко описать эти атрибуты. КМК можно оставить и protected resource, если воспринимать это больше как абстракцию, что мы бэкапим этот ресурс и у него есть точки восстановления. И в зависимости от даты последних точек восстановления можно судить насколько протух/актуален бэкапируемый ресурс в нашем хранилище

RudeZwloki7 replied:
Обсудили, лучше исправить на recovery point
